### PR TITLE
13.1 Validator

### PR DESCRIPTION
**Pull request by xalivan**

### CODE REVIEW

File: src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                    .field(FieldType.DATE.getField())
+                    .build();
+        }
+
+        return null;
+    }
+
+    @Override
+    public Supplier<String> getField(MessageDTO messageDTO) {
+        return () -> convertToString(messageDTO.getDate());
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java:48]: what is this magic? you can explain why we do not use generics?

File: src/main/java/com/speaker/service/validator/message/AbstractMessageCheckValidator.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.speaker.service.validator.message;
+
+import com.speaker.dto.MessageDTO;
+import com.speaker.service.validator.Validator;
+
+import static java.util.Objects.isNull;
+
+public  abstract class AbstractMessageCheckValidator implements Validator<MessageDTO, String> {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/AbstractMessageCheckValidator.java:8]: you have lost all the benefits of polymorphism :(

File: src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public ValidatorError validate(String data) {
+        if (checkOrNull(data)) {
+            return ValidatorError.builder()
+                    .message(FieldType.DATE.getField() + ErrorType.EMPTY.getError())
+                    .field(FieldType.DATE.getField())
+                    .build();
+        }
+        if (splitByT(data).isEmpty()) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java:29]: `splitByT` - bad name and unnecessary method, you can work with Date object for example

File: src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+
+import static java.util.Objects.isNull;
+
+@Component
+public class MessageDataValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+    private final long CURRENT_TIME = System.currentTimeMillis();
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java:19]: this is not the current time!
in this case `CURRENT_TIME` - the time where the bean was created

File: src/main/java/com/speaker/service/MessageServiceImpl.java [hunk 5]
```diff
@@ -22,22 +25,31 @@ public class MessageServiceImpl implements MessageService {
...
     @Override
     public List<Message> getMessagesByAccountId(int accountId) {
         return messageRepository.findAllByAccountId(accountId);
     }
 
-    @Override
-    public Response addMessage(List<MessageDTO> messageDTOs) {
+        @Override
+    public List<ValidatorError> addMessage(List<MessageDTO> messageDTOs) {
+        List<ValidatorError> validatorErrors = messageDTOs.stream()
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/MessageServiceImpl.java:37]: let's do two methods in `FieldValidators`:

1)  `List<ValidatorError> validate(T entity);`
2)  `List<ValidatorError> validate(List<T> enties);`

File: src/main/java/com/speaker/service/MessageServiceImpl.java [hunk 6]
```diff
@@ -22,22 +25,31 @@ public class MessageServiceImpl implements MessageService {
...
+            return validatorErrors;
+        }
         List<Message> messages = messageDTOs.stream()
                 .map(this::convertToMessage)
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
         if (messageDTOs.size() == messages.size()) {
-            return messageRepository.createMessages(messages) ? Response.TRUE : Response.FALSE;
+           messageRepository.createMessages(messages);
+           return List.of(new ValidatorError("messages  created", ""));
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/MessageServiceImpl.java:50]: `new ValidatorError("messages was created", "messageDTOs")`

File: src/main/java/com/speaker/service/MessageServiceImpl.java [hunk 7]
```diff
@@ -22,22 +25,31 @@ public class MessageServiceImpl implements MessageService {
...
         }
-        return Response.FALSE;
+        return List.of(new ValidatorError("messages no created", ""));
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/MessageServiceImpl.java:52]: also missed field name

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.stereotype.Component;
+
+import static java.util.Objects.isNull;
+
+@Component
+public class MessageFromAccountIdValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+
+    @Override
+    public ValidatorError validate(EntityField entityField) {
+        if (isNull(entityField.getField().toString())) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java:19]: you get NPE if a message will be missed

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+            return createValidatorError(ErrorType.EMPTY);
+        }
+        if (!parseToInt(entityField)) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java:22]: Why does it need?

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+            return createValidatorError(ErrorType.NOT_VALID);
+        }
+        if ((int) entityField.getField() < 0) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java:25]: if the value will be 0?

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountNamesValidatorImpl.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+@RequiredArgsConstructor
+public class MessageFromAccountNamesValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+    private final AccountRepository accountRepository;
+
+    @Override
+    public ValidatorError validate(EntityField entityField) {
+        if (isNull(entityField.getField().toString())) {
+            return createValidatorError(ErrorType.EMPTY);
+        }
+        if (splitBySpace(entityField.getField().toString()).isEmpty()) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageFromAccountNamesValidatorImpl.java:27]: extract in variable `entity Field.getField()`

File: src/main/java/com/speaker/service/validator/message/MessageStatusValidatorImpl.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.Objects;
+
+@Component
+public class MessageStatusValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+
+    @Override
+    public ValidatorError validate(EntityField entityField) {
+        if (!Objects.equals(entityField.getField().toString(), Status.NEW.name())) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageStatusValidatorImpl.java:20]: you get NPE
just `Status.NEW.equals(entityField.getField())`

File: src/main/java/com/speaker/service/validator/message/MessageTextValidatorImpl.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+import static java.util.Objects.isNull;
+
+@Component
+public class MessageTextValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+    @Override
+    public ValidatorError validate(EntityField entityField) {
+        if (isNull(entityField.getField().toString())) {
+            return createValidatorError(ErrorType.EMPTY);
+        }
+        if (entityField.getField().equals("")) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageTextValidatorImpl.java:21]: if we will have "  ", what then?

File: src/main/java/com/speaker/service/validator/message/MessageToAccountNamesValidatorImpl.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+import static java.util.Objects.isNull;
+
+@Component
+@RequiredArgsConstructor
+public class MessageToAccountNamesValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+    private final AccountRepository accountRepository;
+
+    @Override
+    public ValidatorError validate(EntityField entityField) {
+        if (isNull(entityField.getField().toString())) {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/validator/message/MessageToAccountNamesValidatorImpl.java:24]: please extract in all places `entity Field.getField()` to variable

File: src/main/java/com/speaker/service/MessageServiceImpl.java (hunk 5 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/MessageServiceImpl.java:37]: Comments accepted, corrections made. The interface is extended

File: src/main/java/com/speaker/service/MessageServiceImpl.java (hunk 6 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/MessageServiceImpl.java:50]: Comments accepted, corrections made. The message is extended

File: src/main/java/com/speaker/service/MessageServiceImpl.java (hunk 7 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/MessageServiceImpl.java:52]: Comments accepted, corrections made. The message is extended

File: src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java (hunk 4 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java:19]: Comments accepted, corrections made. The time is changed
    private final long CURRENT_TIME = LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java (hunk 9 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java:22]: Comments accepted, corrections made. Deleted  if (!parseToInt(entityField)) {}

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java (hunk 10 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageFromAccountIdValidatorImpl.java:25]: Comments accepted, corrections made.    
if ((int) entityField.getField() <= 0)

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountNamesValidatorImpl.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+import static java.util.Objects.isNull;
+
+@Component
+@RequiredArgsConstructor
+public class MessageFromAccountNamesValidatorImpl extends AbstractValidator implements Validator<MessageDTO> {
+    private final AccountRepository accountRepository;
+
+    @Override
+    public ValidatorError validate(EntityField entityField) {
+        if (isNull(entityField.getField().toString())) {
```
[the remark this answers is not in the document]
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageFromAccountNamesValidatorImpl.java:24]: Comments accepted, corrections made.
Object fieldName = entityField.getField();
        if (isNull(fieldName))

File: src/main/java/com/speaker/service/validator/message/MessageFromAccountNamesValidatorImpl.java (hunk 11 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageFromAccountNamesValidatorImpl.java:27]: Comments accepted, corrections made.

File: src/main/java/com/speaker/service/validator/message/MessageStatusValidatorImpl.java (hunk 12 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageStatusValidatorImpl.java:20]: Comments accepted, corrections made.

File: src/main/java/com/speaker/service/validator/message/MessageToAccountNamesValidatorImpl.java (hunk 14 above)
Comment (xalivan) [reply src/main/java/com/speaker/service/validator/message/MessageToAccountNamesValidatorImpl.java:24]: Comments accepted, corrections made.

File: src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java (hunk 4 above)
Comment (otkalenko) [reply src/main/java/com/speaker/service/validator/message/MessageDataValidatorImpl.java]: I mean that it final variable, and fill one time when created bean in spring context. If the progect was started two days ago this variable will have date = `current time - 2 days`. 
Please do not extract this into a final variable, use it in the method.

**Pull request merged**